File: fixtures/ts-sample/src/direct.ts
// cgx-fixture: direct function calls (TypeScript)
// Covers: calls (probable with SCIP, possible without),
//         method calls on typed class instances (certain with type info),
//         static method calls

export function add(a: number, b: number): number {
    return innerAdd(a, b);
}

function innerAdd(a: number, b: number): number {
    return a + b;
}

export function identity<T>(x: T): T {
    return cloneValue(x);
}

function cloneValue<T>(x: T): T {
    return x;
}

export class Counter {
    private value: number;

    constructor(start: number) {
        this.value = start;
    }

    increment(): void {
        this.value = this.addOne(this.value);  // certain: typed receiver
    }

    private addOne(n: number): number {
        return n + 1;
    }

    get(): number {
        return this.value;
    }

    static create(start: number): Counter {
        return new Counter(start);
    }
}

export function chain(n: number): number {
    const a = stepA(n);
    const b = stepB(a);
    return stepC(b);
}

function stepA(n: number): number { return n + 1; }
function stepB(n: number): number { return n * 2; }
function stepC(n: number): number { return n - 3; }

export function useCounter(): number {
    const c = Counter.create(0);  // direct static method call
    c.increment();
    c.increment();
    return c.get();
}

File: fixtures/rust-sample/src/direct.rs
// cgx-fixture: direct calls
// Covers: calls (certain), monomorphized generics (certain),
//         static method calls (certain)

/// Add two integers. Direct call target.
pub fn add(a: i32, b: i32) -> i32 {
    inner_add(a, b)
}

fn inner_add(a: i32, b: i32) -> i32 {
    a + b
}

/// Generic monomorphized call — resolved to certain at each call site.
pub fn identity<T: Clone>(x: T) -> T {
    clone_value(x)
}

fn clone_value<T: Clone>(x: T) -> T {
    x.clone()
}

/// Method on a concrete struct — direct call, certain confidence.
pub struct Counter {
    value: i32,
}

impl Counter {
    pub fn new(start: i32) -> Self {
        Counter { value: start }
    }

    pub fn increment(&mut self) {
        self.value = self.add_one(self.value);
    }

    fn add_one(&self, n: i32) -> i32 {
        n + 1
    }

    pub fn get(&self) -> i32 {
        self.value
    }
}

/// Chained direct calls.
pub fn chain(n: i32) -> i32 {
    let a = step_a(n);
    let b = step_b(a);
    step_c(b)
}

fn step_a(n: i32) -> i32 { n + 1 }
fn step_b(n: i32) -> i32 { n * 2 }
fn step_c(n: i32) -> i32 { n - 3 }
